gh-118362: Fix thread safety around lookups from the type cache in the face of concurrent mutators

DinoV opened the pull request:
Lookups from the type cache can be exposed to some races.  One of those is that `_PyType_Lookup` isn't increfing the result, so the resulting value can become invalid at any point.  This adds a new `_PyType_Fetch` API which does the incref.  Most places are updated to use the new API except for the specializer which isn't safe in free-threaded builds yet.

But there are also issues around when we do a store into the dict and signal that the type is modified.  A race can occur where the old value is removed from the dict and before we invalidate the type version.  This can be seen today w/ the GIL when the value in the dict has a finalizer - accessing the type cache and the type's mapping proxy return inconsistent results.

This fixes these issues by making the update to the dict and the type modified update an atomic operation.  We also capture the previous value in the dictionary and don't free it until after we've made the atomic update.  This basically inlines a much simplified version of `_PyObject_GenericSetAttrWithDict` into `type_setattro` which can lock mutation against types as well as against the type's dict.  

When updating we first clear the type version so concurrent reads of the type cache won't hit.  We then replace the existing value and finish the type modified process.

<!-- gh-issue-number: gh-118362 -->
* Issue: gh-118362
<!-- /gh-issue-number -->

DinoV committed:
Add _PyType_Fetch and use incref before setting attribute on type
Makes setting an attribute on a class and signaling type modified atomic
Avoid adding re-entrancy exposing the type cache in an inconsistent state by decrefing after type is updated

colesbury commented on Include/cpython/object.h:
```diff
@@ -275,6 +275,7 @@ typedef struct _heaptypeobject {
 
 PyAPI_FUNC(const char *) _PyType_Name(PyTypeObject *);
 PyAPI_FUNC(PyObject *) _PyType_Lookup(PyTypeObject *, PyObject *);
+PyAPI_FUNC(PyObject *) _PyType_Fetch(PyTypeObject *, PyObject *);
```
I think the naming convention that's evolved in 3.13 is that function variants that return a strong reference have the `Ref` suffix (like `PyDict_GetItemRef`, `PyList_GetItemRef`, `PyWeakref_GetRef`) so maybe `_PyType_LookupRef`?

colesbury commented on Objects/dictobject.c:
```diff
@@ -2317,6 +2317,37 @@ PyDict_GetItemRef(PyObject *op, PyObject *key, PyObject **result)
+        }
+    }
+
+    PyDictObject*mp = (PyDictObject *)op;
```
nit: space in `PyDictObject*mp`

colesbury commented on Objects/dictobject.c:
```diff
@@ -2317,6 +2317,37 @@ PyDict_GetItemRef(PyObject *op, PyObject *key, PyObject **result)
+_PyDict_GetItemRef_LockHeld(PyObject *op, PyObject *key, PyObject **result)
+{
+    ASSERT_DICT_LOCKED(op);
+    assert(PyUnicode_CheckExact(key));
```
This seems a bit strange in that the unicode requirement is not indicated by the function name and not enforced by the similarly named `_PyDict_SetItem_LockHeld`.

DinoV replied:
I mainly didn't want a ridiculously long name and it's similar to `_PyDict_GetItemWithError` which doesn't call it out, but I'll go ahead and throw `Unicode` in its name :)

colesbury commented on Objects/object.c:
```diff
@@ -1745,16 +1743,10 @@ _PyObject_GenericSetAttrWithDict(PyObject *obj, PyObject *name,
-                         "'%.100s' object has no attribute '%U'",
-                         tp->tp_name, name);
-        }
+        assert(!PyType_IsSubtype(tp, &PyType_Type));
```
I don't understand this assert: why can't `tp` be a `PyType_Type` subtype?

DinoV replied:
This is to make sure that removing the type specific special case here is safe.  I'll actually hoist this up to the top of the function instead of burying it here as we don't want to use this on types.

DinoV replied:
Moving this actually caused this to get hit in ctypes...  But I think there's a good reason to not do this - the type cache isn't properly getting invalidated if using this directly.

colesbury commented on Objects/typeobject.c:
```diff
@@ -849,6 +861,43 @@ PyType_Unwatch(int watcher_id, PyObject* obj)
+        }
+    }
+
+    _Py_atomic_store_uint32_release(&type->tp_version_tag, 0); /* 0 is not a valid version tag */
```
nit: line length (maybe move the comment above?)

colesbury commented on Objects/typeobject.c:
```diff
@@ -882,7 +931,7 @@ type_modified_unlocked(PyTypeObject *type)
                 continue;
             }
-            PyType_Modified(subclass);
+            type_modified_unlocked(subclass);
```
👍

colesbury commented on Objects/typeobject.c:
```diff
@@ -5161,6 +5213,14 @@ _PyType_Lookup(PyTypeObject *type, PyObject *name)
+{
+    PyObject *res = _PyType_Fetch(type, name);
+    Py_XDECREF(res);
+    return res;
```
This strategy worries me a bit. I'd be more comfortable if the `_PyType_Lookup` doesn't incref+decref the returned object.

colesbury commented on Objects/typeobject.c:
```diff
@@ -5341,49 +5399,107 @@ type_setattro(PyObject *self, PyObject *name, PyObject *value)
-        if (PyUnicode_CheckExact(name)) {
-            Py_INCREF(name);
+    if (!PyUnicode_Check(name)) {
+        PyErr_Format(PyExc_TypeError,
+                     "attribute name must be string, not '%.200s'",
+                     Py_TYPE(name)->tp_name);
```
Just checking my understanding: this was already enforced in `_PyObject_GenericSetAttrWithDict`, right? So it's not a new restriction, it's just that the error is raised earlier in the code path.

colesbury commented on Objects/typeobject.c:
```diff
@@ -5341,49 +5399,107 @@ type_setattro(PyObject *self, PyObject *name, PyObject *value)
+    }
+
+    PyTypeObject *metatype = Py_TYPE(type);
+    assert(!_PyType_HasFeature(metatype, Py_TPFLAGS_INLINE_VALUES));
+    assert(!_PyType_HasFeature(metatype, Py_TPFLAGS_MANAGED_DICT));
```
How can we be sure the metatype doesn't use a managed dict (or even a different getattr implementation?)

colesbury commented on Objects/typeobject.c:
```diff
@@ -5341,49 +5399,107 @@ type_setattro(PyObject *self, PyObject *name, PyObject *value)
+    assert(!_PyType_HasFeature(metatype, Py_TPFLAGS_INLINE_VALUES));
+    assert(!_PyType_HasFeature(metatype, Py_TPFLAGS_MANAGED_DICT));
+
+    PyObject *old_value;
```
I'd initialize `old_value` to `NULL` here instead of on line 5437 below. Although, I don't think it's necessary in the current implementation, I think it'll make the code more robust to future changes.

colesbury commented on Objects/typeobject.c:
```diff
@@ -5341,49 +5399,107 @@ type_setattro(PyObject *self, PyObject *name, PyObject *value)
+        BEGIN_TYPE_LOCK();
+        dict = type->tp_dict;
+        if (dict == NULL) {
+            dict = type->tp_dict = PyDict_New();
```
I think we need to handle the case where `PyDict_New()` returns NULL.

colesbury commented on Objects/typeobject.c:
```diff
@@ -5341,49 +5399,107 @@ type_setattro(PyObject *self, PyObject *name, PyObject *value)
-           recursing into subclasses. */
-        type_modified_unlocked(type);
+#ifdef Py_GIL_DISABLED
+    type_modification_starting_unlocked(type);
```
Might be worth a comment the purpose of this.

colesbury commented on Objects/typeobject.c:
```diff
@@ -5341,49 +5399,107 @@ type_setattro(PyObject *self, PyObject *name, PyObject *value)
+    type_modification_starting_unlocked(type);
+#endif
+
+    res = _PyDict_SetItem_LockHeld((PyDictObject *)dict, name, value);
```
Minor, but it's a little strange that `_PyDict_GetItemRef_LockHeld` takes a `PyObject*` but `_PyDict_SetItem_LockHeld` takes a `PyDictObject*`

DinoV replied:
I had initially been going with `PyDictObject` reflecting a recent comment from @markshannon that we should type things correctly, but there are other places where things aren't typed correctly (like `tp_dict` which is just a `PyObject`) that introduced extra casts.  I'll go ahead and add the casts.

colesbury commented on Objects/typeobject.c:
```diff
@@ -5341,49 +5399,107 @@ type_setattro(PyObject *self, PyObject *name, PyObject *value)
+    // the locks.
+    BEGIN_TYPE_DICT_LOCK(dict);
+
+    if (_PyDict_GetItemRef_LockHeld(dict, name, &old_value) < 0) {
```
We're doing both "preserve the old value" and "invalidate the cache early" strategies. Is preserving the old value necessary for safety?

(To be clear: I'm not opposed to this; I'm just trying to understand the implementation and constraints).

DinoV replied:
This is necessary to prevent any re-entrancy during the update (when the existing value is replaced and its finalizer runs). The problem with clearing the version tag but not clearing `Py_TPFLAGS_VALID_VERSION_TAG` is that we're in a pretty funky state.  If we start running Python code again and hit `assign_version_tag` anywhere we'll happily think we have a version tag assigned even though our version is invalid.  That means we can do a lookup against that and get a cache hit.  Ultimately that breaks in the new test in `test_class`.

By guaranteeing that we have no re-entrancy we're not exposing this funky intermediate state.  And we need to have this funky intermediate state so that we can both not have any cache hits as well as being able to know what types we need to clear later when we do the full `type_modified_unlocked`.

The alternative here I think would be to clear both the type version tag and the has version flag.  That would also require your originally proposed flag to disable the type cache.  But the challenging thing there is then finishing the 2nd half of the `type_modified_unlocked` - we need to remember all of the types that we've cleared, but we lost the state that's tracking that (`Py_TPFLAGS_VALID_VERSION_TAG`).  So that would require either another flag to say we're in the process of clearing the tags or we'd need to store all of the types that we're clearing in a list.

But if we avoid the re-entrancy then everything's fine, and assigning a new version number to the types is protected by the type lock.

DinoV replied:
Oh, also I wanted to assert that we're never caching w/ a 0 version tag.  And that works fine everywhere except for in some tests for specialization which are using `_testcapi` to assign a 0 version tag and say it's valid :(

colesbury commented on Objects/typeobject.c:
```diff
@@ -5341,49 +5399,107 @@ type_setattro(PyObject *self, PyObject *name, PyObject *value)
-        assert(_PyType_CheckConsistency(type));
     }
-    END_TYPE_LOCK()
+    else if (PyErr_ExceptionMatches(PyExc_KeyError)) {
```
This is because we're not calling `_PyObject_GenericSetAttrWithDict` directly, right?

Should we refactor the common code out?

DinoV replied:
Partly, I think the check and the custom error message are an improvement being here, it was weird having the special case for type objects in the generic function.  There's the logic for setting name and obj which I'll expose as common instead of being a static function in object.

colesbury commented on Lib/test/test_free_threading/test_type.py:
```diff
@@ -0,0 +1,112 @@
+NTHREADS = 6
+BOTTOM = 0
+TOP = 1000
+ITERS = 100
```
How long does this test take to run in a debug build? The `test_free_threading` tests are a bit slow right now (~1 min in debug builds on Linux, I think). I think we should aim for the default to be fast, even if that means it's less likely to catch bugs. (We can add flag or option to run for longer).

DinoV replied:
Currently 9 seconds total (`test_attr_cache` is actually pretty fast, the other two are slower).  I'll reduce the loop length in the other 2 and which gets it down to 2-3 seconds.

colesbury reviewed (commented):
Some questions comments below. Biggest questions is that now that we are accessing the `dict` directly instead of through `_PyObject_GenericSetAttrWithDict`, how do we know that the other cases handled by `_PyObject_GenericSetAttrWithDict` are not necessary in `_Py_type_getattro`?

Also, I think this merits a NEWS entry.

DinoV committed:
Rename _PyType_Fetch -> _PyType_LookupRef
Fix some formatting
Add news blurb
Use PyDictObject * more
Rename _PyDict_GetItemRef_LockHeld to _PyDict_GetItemRef_Unicode_LockHeld
Reduce iterations in test
Expose _PyObject_SetAttributeErrorContext for attaching context

DinoV committed:
ctypes should use base implemetantation for setting attribute

colesbury commented on Include/internal/pycore_object.h:
```diff
@@ -658,6 +658,7 @@ extern PyObject *_PyType_NewManagedObject(PyTypeObject *type);
 extern PyTypeObject* _PyType_CalculateMetaclass(PyTypeObject *, PyObject *);
 extern PyObject* _PyType_GetDocFromInternalDoc(const char *, const char *);
 extern PyObject* _PyType_GetTextSignatureFromInternalDoc(const char *, const char *, int);
+extern int _PyObject_SetAttributeErrorContext(PyObject* v, PyObject* name);
```
```suggestion
extern int _PyObject_SetAttributeErrorContext(PyObject *v, PyObject *name);
```

colesbury commented on Objects/typeobject.c:
```diff
@@ -4981,7 +5026,10 @@ find_name_in_mro(PyTypeObject *type, PyObject *name, int *error)
         PyObject *dict = lookup_tp_dict(_PyType_CAST(base));
         assert(dict && PyDict_Check(dict));
-        res = _PyDict_GetItem_KnownHash(dict, name, hash);
+        if (_PyDict_GetItemRef_KnownHash((PyDictObject *)dict, name, hash, &res) < 0) {
```
I think the `PyErr_Occurred()` check below can be removed now.

colesbury reviewed (approved):
LGTM with a few minor issues:

* We need to handle the `PyDict_New()` returning `NULL` case in `type_setattro`
* Minor formatting in `_PyObject_SetAttributeErrorContext`
* `PyErr_Occurred()` case is no longer need in `find_name_in_mro`

DinoV committed:
Handle failure to create dict, fix formatting, remove extra pyerr_occurred

colesbury commented on Lib/test/test_free_threading/test_type.py:
```diff
@@ -0,0 +1,112 @@
+class A:
+    attr = 1
+
+@unittest.skipIf(is_wasi, "WASI has no threads.")
```
Oh, probably better to use `@threading_helper.requires_working_threading()`

mhsmith replied:
iOS and Android do both have working threading, but they don't support subprocesses. In theory `multiprocessing.dummy.Pool` should work on these platforms, but in practice it doesn't because it imports too much of the main multiprocessing module.

The simplest solution is probably to use `concurrent.futures.ThreadPoolExecutor` instead.
